### PR TITLE
Add new XTENSA_ARCH_HAS_HIFIn

### PR DESCRIPTION
**Pull request by marc-hb**
WIP.

On-going discussions:
- https://github.com/thesofproject/sof/issues/8697
- https://github.com/thesofproject/sof/pull/8720
- https://github.com/thesofproject/sof/pull/8733

### CODE REVIEW

**Review by andyross — COMMENTED** [review]
Seems reasonable.  FWIW, while I'm sure there was indeed a "1" and "2" in the depths of history, all Xtensa hardware on which Zephyr is running or planned has one of the HiFi3/4/5 units.   Cadence offers a lot of coprocessor IP, and we don't need to declare all of it.

Also given that this recapitulates the various XCHAL_HAVE_* preprocessor macros from core-isa.h (which I understand isn't quite enough to discriminate at build time given that we may have other concerns like binary compatibility or toolchain choice that affect how we generate code), it would be a good to have a header in the arch somewhere that guarantees that these settings are compatible.  That is, if we have CONFIG_XTENSA_ARCH_HAS_HIFI4=y, then XCHAL_HAVE_HIFI4_VFPU must be defined to 1 in the core-isa.h we're getting off the include path.  If it isn't, we're either building for the wrong device or the toolchain or HAL is giving us the wrong headers.

**Comment by marc-hb** [comment]
> Cadence offers a lot of coprocessor IP, and we don't need to declare all of it.

Good point, I'll remove 1 and 2. I found them in some old Intel toolchains but no point adding them to Zephyr if they are not in use here. They could easily be added later if needed (= never).


> ... recapitulates the various XCHAL_HAVE_* preprocessor macros from core-isa.h (which I understand isn't quite enough to discriminate at build time given that we may have other concerns like binary compatibility or toolchain choice that affect how we generate code)

To be really honest  I would love a more formal, short-list of high-level requirements in a single place that explain why I'm doing this. I think most of them can be found in the review comments of https://github.com/thesofproject/sof/pull/8720, https://github.com/thesofproject/sof/pull/8682 and a couple others but that's inconvenient and unreliable.

Or maybe @peter-mitsis will provide a different set of high-level requirements for saving HIFI registers that will magically need the same implementation and we can save that documentation effort :-)

> That is, if we have CONFIG_XTENSA_ARCH_HAS_HIFI4=y, then XCHAL_HAVE_HIFI4_VFPU must be defined to 1 in the core-isa.h we're getting off the include path. If it isn't, we're either building for the wrong device or the toolchain or HAL is giving us the wrong headers.

Automatically checking at build time that this duplication does not diverge; good idea, thanks! Note one of the important SOF requirement was NOT to `#include <core-isa.h>` (https://github.com/thesofproject/sof/pull/8682#issuecomment-1877306254, ...) but maybe in Zephyr it's OK.

**Comment by peter-mitsis** [comment]
Thanks for putting this together! Although I am working on saving/restoring the HiFi registers, it has not been 100% clear to me yet which generation of the HiFi coprocessors would need to be supported. This really helps clarify things.

If we can get those compliance checks fixed, I would really like to see this merged so that I can build off of it.

File: soc/xtensa/intel_adsp/ace/Kconfig.series [hunk 1]
```diff
@@ -5,7 +5,7 @@ config SOC_SERIES_INTEL_ACE
 	bool "Intel ACE"
 	select SOC_FAMILY_INTEL_ADSP
 	select XTENSA
-	select XTENSA_HAL if (("$(ZEPHYR_TOOLCHAIN_VARIANT)" != "xcc") && ("$(ZEPHYR_TOOLCHAIN_VARIANT)" != "xt-clang"))
+	select XTENSA_HAL if ! XTENSA_TOOLCHAIN_HAS_HIFI
```
Comment (btian1) [inline soc/xtensa/intel_adsp/ace/Kconfig.series:8]: ! XTENSA_TOOLCHAIN_HAS_HIFI --> !XTENSA_TOOLCHAIN_HAS_HIFI

File: arch/xtensa/Kconfig [hunk 2]
```diff
@@ -93,6 +93,32 @@ config XTENSA_NUM_SPIN_RELAX_NOPS
...
+       help
+         HiFi2 Audio Engine
+
+config XTENSA_ARCH_HAS_HIFI3
+       bool
+       help
+         HiFi3 Audio Engine
+
+config XTENSA_ARCH_HAS_HIFI4
+       bool
```
Comment (btian1) [inline arch/xtensa/Kconfig:113]: can we add more to make this be true or false, like below:

default y if SOC_INTEL_ACE15_MTPM || SOC_INTEL_ACE20_LNL

or you want to add similar in other place?
I mean we need have a place to enable these config.

File: soc/xtensa/intel_adsp/ace/Kconfig.series [hunk 3]
```diff
@@ -6,6 +6,8 @@ config SOC_SERIES_INTEL_ACE
 	select SOC_FAMILY_INTEL_ADSP
 	select XTENSA
 	select XTENSA_HAL if ! XTENSA_TOOLCHAIN_HAS_HIFI
+	select XTENSA_ARCH_HAS_HIFI3 if XTENSA_TOOLCHAIN_HAS_HIFI
+	select XTENSA_ARCH_HAS_HIFI4 if XTENSA_TOOLCHAIN_HAS_HIFI
```
Comment (btian1) [inline soc/xtensa/intel_adsp/ace/Kconfig.series:10]: there is no difference for HIFI3 and HIFI4, can we keep one only for ACE platform?

File: arch/xtensa/Kconfig (hunk 2 above)
Comment (marc-hb) [reply arch/xtensa/Kconfig:113]: No `default` is supposed to be here https://docs.zephyrproject.org/latest/build/kconfig/tips.html#using-select-for-helper-symbols

Comment (btian1) [reply arch/xtensa/Kconfig:113]: @andyross @dbaluta @kv2019i Can we say that all SOF users are now go with zephyr even for IPC3? 
i.e, all existing SOF users are depend on zephyr?

Comment (dbaluta) [reply arch/xtensa/Kconfig:113]: > @andyross @dbaluta @kv2019i Can we say that all SOF users are now go with zephyr even for IPC3?
> i.e, all existing SOF users are depend on zephyr(no matter IPC3/IPC4)?
> AMD, MTK, NXP are all depend on zephyr now?

For NXP, all platforms are supported both with Zephyr and XTOS  (IPC3-only, so far).

Comment (btian1) [reply arch/xtensa/Kconfig:113]: Thanks, Daniel, so this year means 2024, right?
I think it is quite clear now: at least, NXP still have XTOS based IPC3 build, which means we still can't fully depend on zephyr to pass HIFIn information to SOF.

**Comment by marc-hb** [comment]
The much smaller approach https://github.com/thesofproject/sof/pull/8787 has just been merged.

Long story short, we're not duplicating every `core-isa.h` file in Kconfig for now. Maybe later. We would have had to duplicate each `core-isa.h`  TWICE: once for Zephyr and once for legacy, pre-Zephyr XTOS.

@peter-mitsis I'm closing this PR but I'm not deleting the branch in case you want to re-use some of it for saving and restoring registers.

**Pull request closed without merging**